### PR TITLE
feat(core/beforeSave): support modifications before save

### PR DESCRIPTION
**Pull request by marcoscaceres**
Adds a `beforeSave` configuration option, which takes an array of sync JS functions. Unfortunately, due to a limitation of pubsubhub, we can't support async functions yet. But that should be ok for now.

### CODE REVIEW

File: src/core/beforeSave.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+      let pos = 0;
+      for (const fn of conf.beforeSave) {
+        try {
+          fn(ownerDocument);
+        } catch (err) {
+          const nameOrPosition = `\`${fn.name}\`` || `at position ${pos}`;
+          const msg = docLink`Function ${nameOrPosition}\` threw an error during processing of ${"[beforeSave]"}.`;
+          const hint = "See developer console.";
+          showError(msg, name, { hint });
+          console.error(err);
```
Comment (sidvishnoi) [inline src/core/beforeSave.js:32]: I think we support passing `{error}` in `showError`? I might be wrong. We should support that if not.

File: src/core/beforeSave.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    return;
+  }
+
+  sub(
+    "beforesave",
+    ({ ownerDocument }) => {
+      let pos = 0;
+      for (const fn of conf.beforeSave) {
+        try {
+          fn(ownerDocument);
```
Comment (sidvishnoi) [inline src/core/beforeSave.js:26]: Should it also have access to `conf`?

File: src/core/beforeSave.js [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    )
+  ) {
+    const msg = docLink`${"[beforeSave]"} configuration option must be an array of synchronous JS functions.`;
+    showError(msg, name);
+    return;
+  }
+
+  sub(
+    "beforesave",
+    ({ ownerDocument }) => {
```
Comment (sidvishnoi) [inline src/core/beforeSave.js:22]: Nit: let's move this into a named function for clarity.

File: tests/spec/core/beforeSave-spec.html [hunk 4]
```diff
@@ -0,0 +1,32 @@
+<!DOCTYPE html>
+<script class="remove">
+  const modifier1 = exportDoc => {
```
Comment (sidvishnoi) [inline tests/spec/core/beforeSave-spec.html:3]: Nit:
```suggestion
  const modifier1 = exportedDoc => {
```

File: tests/spec/core/beforeSave-spec.html [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    exportDoc.body.appendChild(p);
+  };
+  const respecConfig = {
+    specStatus: "unofficial",
+    shortName: "i",
+    editors: [
+      {
+        name: "Foo",
+        url: "https://foo.com/",
+      },
```
Comment (sidvishnoi) [inline tests/spec/core/beforeSave-spec.html:22]:
```suggestion
      { name: "Foo", url: "https://foo.com/" },
```
(_OCD_)

File: tests/spec/core/beforeSave-spec.js [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    const p2 = exportedDoc.querySelector("#p2");
+    expect(p1.nextElementSibling).toBe(p2);
+  });
+
+  it("complains if it's not passed an array", async () => {
+    const ops = makeStandardOps({ beforeSave: "not a array" });
+    const doc = await makeRSDoc(ops);
+    const errors = doc.respec.errors.filter(
+      err => err.plugin === "core/beforeSave"
+    );
```
Comment (sidvishnoi) [inline tests/spec/core/beforeSave-spec.js:31]: Nit: can move this repeated code block into a function `getErrors(doc)` to the parent `describe` block.

**Review by sidvishnoi — COMMENTED** [review]
Looks good. What prompted for this feature though?
(In future, I would love if we can have https://github.com/w3c/respec/pull/3320.)

File: src/core/beforeSave.js (hunk 2 above)
Comment (marcoscaceres) [reply src/core/beforeSave.js:26]: I don't think so, tbh. The conf is already global and we want to discourage people poking at the private stuff.

File: src/core/beforeSave.js (hunk 1 above)
Comment (marcoscaceres) [reply src/core/beforeSave.js:32]: ah, let me check...

Comment (sidvishnoi) [reply src/core/beforeSave.js]: We don't. We can however use `Error.cause`. Let's do that as follow up.

File: profiles/w3c.js [hunk 7]
```diff
@@ -59,6 +59,7 @@ const modules = [
   import("../src/core/anchor-expander.js"),
   import("../src/core/custom-elements/index.js"),
   import("../src/core/web-monetization.js"),
+  import("../src/core/beforeSave.js"),
```
Comment (sidvishnoi) [inline profiles/w3c.js:62]: Let's follow the file naming conventions we already have (i.e. `before-save.js`)

Comment (sidvishnoi) [inline profiles/w3c.js:62]: Might want to add it to other profiles too as it's a core feature.

Comment (marcoscaceres) [reply profiles/w3c.js:62]: I'll let there other folks do that, as it's up to them to maintain their own profiles.

File: tests/spec/core/beforeSave-spec.js (hunk 6 above)
Comment (marcoscaceres) [reply tests/spec/core/beforeSave-spec.js]: Added UIMessageFilters, and errorFilters, and warningFilters exports... wdyt?

File: tests/spec/SpecHelper.js [hunk 8]
```diff
@@ -56,6 +56,34 @@ export function makeRSDoc(opts, src, style = "") {
...
+    this.type = type;
+  }
+  /**
+   * @param {string} pluginName
+   * @returns (Document) => Array<RespecError>
+   */
+  filter(pluginName) {
+    if (this.has(pluginName)) {
+      return this.get(pluginName);
+    }
```
Comment (sidvishnoi) [inline tests/spec/SpecHelper.js:77]: Not following why we might want a Map here.

Comment (marcoscaceres) [reply tests/spec/SpecHelper.js:77]: You always need a map :) ... but I'm thinking I'll just make it internal, not extension.

File: src/core/beforeSave.js (hunk 1 above)
Comment (marcoscaceres) [reply src/core/beforeSave.js]: Yeah, good idea... been meaning to play with `.cause`

**Comment by marcoscaceres** [comment]
Going ahead with this one, as I need `UIMessageFilters` for another patch I'm working on. 

But we can revise UIMessageFilters there, but hopefully it's all fine now.

**Pull request merged**